src/soc/ibm/power9/istep_8_10.c

```diff
@@ -0,0 +1,3097 @@
+//         {
+//             p9_io_xbus_scominit(l_XbusConnection.first, l_XbusConnection.second, group);
+//         }
+//     }
+// }
+
+// void p9_io_xbus_scominit(
+//     const fapi2::Target<fapi2::TARGET_TYPE_XBUS> &i_target,
+//     const fapi2::Target<fapi2::TARGET_TYPE_XBUS> &i_connected_target,
+//     const uint8_t group)
+// {
+//     const uint8_t LANE_00 = 0;
+//     fapi2::ReturnCode rc = fapi2::FAPI2_RC_SUCCESS;
+//     // get system target
+//     const fapi2::Target<fapi2::TARGET_TYPE_SYSTEM> l_system_target;
+//     // get proc chip to pass for EC level checks in procedure
+//     fapi2::Target<fapi2::TARGET_TYPE_PROC_CHIP> l_proc = i_target.getParent<fapi2::TARGET_TYPE_PROC_CHIP>();
+//     // assert IO reset to power-up bus endpoint logic
+//     // read-modify-write, set single reset bit (HW auto-clears)
+//     // on writeback
+//     EDIP_RX_IORESET[i_target, group, LANE_00] = 1;
+//     EDIP_RX_IORESET[i_connected_target, group, LANE_00] = 1;
+//     wait_us(50, false);
+//     EDIP_TX_IORESET[i_target, group, LANE_00] = 1;
+//     EDIP_TX_IORESET[i_connected_target, group, LANE_00] = 1;
+//     wait_us(50, false);
+//     // Set rx master/slave attribute prior to calling the scominit procedures.
+//     // The scominit procedure will reference the attribute to set the register field.
+//     set_rx_master_mode(i_target, i_connected_target);
+//     // Set Msb Swap based upon attribute data.
+//     set_msb_swap(i_target, group);
+//     set_msb_swap(i_connected_target, group);
+//     switch (group)
+//     {
+//     case ENUM_ATTR_XBUS_GROUP_0:
+//         p9_xbus_g0_scom(i_target, l_system_target, l_proc);
+//         p9_xbus_g0_scom(i_connected_target, l_system_target, l_proc);
+//         break;
+//     case ENUM_ATTR_XBUS_GROUP_1:
+//         p9_xbus_g1_scom(i_target, l_system_target, l_proc);
+//         p9_xbus_g1_scom(i_connected_target, l_system_target, l_proc);
+//         break;
+//     }
+
+//     fapi2::buffer<uint64_t> l_scom_data;
+//     l_scom_data = l_proc[PU_PB_CENT_SM0_PB_CENT_FIR_REG];
+//     if(!l_scom_data.getBit<13>())
+//     {
+
+//         i_target[XBUS_FIR_ACTION0_REG] = XBUS_PHY_FIR_ACTION0;
+//         i_target[XBUS_FIR_ACTION1_REG] = XBUS_PHY_FIR_ACTION1;
+//         i_target[XBUS_FIR_MASK_REG] = XBUS_PHY_FIR_MASK;
+//     }
+// }
+
+// void set_msb_swap(/*const Target<fapi2::TARGET_TYPE_XBUS>*/uint64_t i_tgt, const uint8_t i_grp)
+// {
+//     const uint8_t lane0   = 0;
+//     uint8_t tx_msb_swap = 0x0;
+//     uint8_t field_data  = 0x0;
+
+//     // Retrieve msb swap attribute
+//     tx_msb_swap = ATTR_EI_BUS_TX_MSBSWAP[i_tgt];
+
+//     if(i_grp == ENUM_ATTR_XBUS_GROUP_0 && (tx_msb_swap & ENUM_ATTR_EI_BUS_TX_MSBSWAP_GROUP_0_SWAP)
+//     || i_grp == ENUM_ATTR_XBUS_GROUP_1 && (tx_msb_swap & ENUM_ATTR_EI_BUS_TX_MSBSWAP_GROUP_1_SWAP))
+//     {
+//             field_data = 1;
+//     }
+//     EDIP_TX_MSBSWAP[i_tgt, i_grp, lane0] = field_data;
+// }
+
+void p9_xbus_g0_scom(
```

Comment:
Seeing the mess with SCOM RMW (read modify write) which is happening here, I would simply implement SCOM access helpers, like:
- scom_and
- scom_or
- scom_and_or
To get it done in oneliners instead of 3-4 lines per register modifications

Reply:
I have added the functions and replaced code here with oneliners.